include/ruby/internal/fl_type.h

```diff
@@ -941,21 +941,8 @@ RB_OBJ_FREEZE_RAW(VALUE obj)
     RB_FL_SET_RAW(obj, RUBY_FL_FREEZE);
 }
 
-/**
- * Prevents further modifications to the given object.  ::rb_eFrozenError shall
- * be raised if modification is attempted.
- *
- * @param[out]  x  Object in question.
- */
-static inline void
-rb_obj_freeze_inline(VALUE x)
-{
-    if (RB_FL_ABLE(x)) {
-        RB_OBJ_FREEZE_RAW(x);
-        if (RBASIC_CLASS(x) && !(RBASIC(x)->flags & RUBY_FL_SINGLETON)) {
-            rb_freeze_singleton_class(x);
-        }
-    }
-}
+RUBY_SYMBOL_EXPORT_BEGIN
+void rb_obj_freeze_inline(VALUE obj);
```

Comment:
inline function is used for performance but is it hard to keep inline?

Reply:
It seems like the idea is to avoid exporting `rb_shape_transition_shape_frozen` as C API.

I think it's okay to make it a non-inline function for C API users even while the `_inline` name becomes a lie, but as Koichi said, it's worth checking if it impacts the interpreter's performance. If it does, you could define it as an inline function in an internal header, and use it as the implementation of (non-inline) `rb_obj_freeze_inline`.

Reply:
Thanks for flagging this. Yes, we removed the inline because we didn't want to export the shape transition function. 

I added [a new benchmark for measuring freezing objects](https://github.com/ruby/ruby/pull/6419).

Running this benchmark, it looks like there's no performance change on freezing objects:

```
compare-ruby: ruby 3.2.0dev (2022-09-20T13:25:05Z master b3d8dddee7) [arm64-darwin21]
built-ruby: ruby 3.2.0dev (2022-09-21T21:33:46Z object-shapes-prot.. d3003b71e3) [arm64-darwin21]
# Iteration per second (i/s)

|               |compare-ruby|built-ruby|
|:--------------|-----------:|---------:|
|vm_freeze_obj  |     364.956|   364.985|
|               |           -|     1.00x|
```